### ISSUE TITLE
Rewrite legacy tests to a more well-documented format using lua/busted

### ISSUE THREAD

**Issue by tarruda** [neovim/neovim#1286]
There's a bunch of tests in src/nvim/testdir which are very hard to read and don't exercise real user interaction(once in a broken PR, no user input was being recognized and yet all tests were passing)

All those tests should be rewritten in a more documented format using lua/busted, where functions are available to communicate with nvim through msgpack-rpc, the same mechanism UIs will use(so it will be very close to a real user interacting with nvim).

Here's a real example, which was previously located at `src/nvim/testdir/test2.in`:

```
This is a test if a URL is recognized by "gf", with the cursor before and
after the "://".  Also test ":\\".

STARTTEST
:so small.vim
/^first
/tmp
:call append(0, expand("<cfile>"))
/^second
/URL
:call append(1, expand("<cfile>"))
:if has("ebcdic")
: set isf=@,240-249,/,.,-,_,+,,,$,:,~,\
:else
: set isf=@,48-57,/,.,-,_,+,,,$,:,~,\
:endif
/^third
/name
:call append(2, expand("<cfile>"))
/^fourth
/URL
:call append(3, expand("<cfile>"))
5GdG:wq! test.out
ENDTEST
first test for URL://machine.name/tmp/vimtest2a and other text
second test for URL://machine.name/tmp/vimtest2b. And other text
third test for URL:\\machine.name\vimtest2c and other text
fourth test for URL:\\machine.name\tmp\vimtest2d, and other text
```

It was converted to `test/functional/legacy/002_filename_recognition_spec.lua`(the _spec suffix is necessary for busted to find the test)

```lua
-- Test if URLs are recognized as filenames by commands such as "gf". Here
-- we'll use `expand("<cfile>")` since "gf" would need to open the file.

local helpers = require('test.functional.helpers')
local clear, feed, insert = helpers.clear, helpers.feed, helpers.insert
local execute, expect = helpers.execute, helpers.expect

describe('filename recognition', function()
  setup(clear)

  it('is working', function()
    -- insert some lines containing URLs
    insert([[
      first test for URL://machine.name/tmp/vimtest2a and other text
      second test for URL://machine.name/tmp/vimtest2b. And other text
      third test for URL:\\machine.name\vimtest2c and other text
      fourth test for URL:\\machine.name\tmp\vimtest2d, and other text]])

    -- Go to the first URL and append it to the beginning
    execute('/^first', '/tmp', 'call append(0, expand("<cfile>"))')

    -- Repeat for the second URL
    -- this time, navigate to the word "URL" instead of "tmp"
    execute('/^second', '/URL', 'call append(1, expand("<cfile>"))')

    -- Repeat for the remaining URLs. This time, the 'isfname' option must be
    -- set to allow '\' in filenames
    execute('set isf=@,48-57,/,.,-,_,+,,,$,:,~,\\')
    execute('/^third', '/name', 'call append(2, expand("<cfile>"))')
    execute('/^fourth', '/URL', 'call append(3, expand("<cfile>"))')

    -- Delete the initial text, which now starts at line 5
    feed('5GdG')

    -- The buffer should now contain:
    expect([[
      URL://machine.name/tmp/vimtest2a
      URL://machine.name/tmp/vimtest2b
      URL:\\machine.name\vimtest2c
      URL:\\machine.name\tmp\vimtest2d]])
  end)
end)
```

Some remarks about the above script:

- the `feed` function can be used to send input to nvim. Special sequences such as `<esc>` or `<cr>` are properly translated.(`helpers` also has a `rawfeed` function which doesn't translate anything)
- the `insert`function wraps `rawfeed` with `i` and `<esc>`(can be used to enter text into the current buffer)
- the `execute` function wraps `rawfeed` with `:` and `<cr>`(it can be used for command-line input)
- the `expect` function takes a string and compares with the current buffer. If it's different, the test will fail
- `setup(clear)` is telling busted(the test framework) to call the `clear` function before running this suite. The `clear` function will perform a bunch of resets to bring nvim as close as possible to it's initial state. Some tests may require a 100% fresh nvim instance, so `setup(restart)` can be used for those cases, though `clear` is usually enough(and faster too because it doesn't require respawning the process). 

Something else is worth mentioning: all strings have common indent across lines removed, so

```lua
    expect([[
      URL://machine.name/tmp/vimtest2a
      URL://machine.name/tmp/vimtest2b
      URL:\\machine.name\vimtest2c
      URL:\\machine.name\tmp\vimtest2d]])
 ```

is the same as

```lua
    expect([[URL://machine.name/tmp/vimtest2a
URL://machine.name/tmp/vimtest2b
URL:\\machine.name\vimtest2c
URL:\\machine.name\tmp\vimtest2d]])
```

**Comment by justinmk** [comment]
## [Checklist for migrating legacy tests](https://github.com/neovim/neovim/wiki/Unit-tests#checklist-for-migrating-legacy-tests)

**Comment by aharris88** [comment]
+1

**Comment by lucc** [comment]
For reference: We had a discussion how to handle aux files when converting legacy tests in #2722. The result was merged in #2932.

**Comment by justinmk** [comment]
After @lucc's existing PRs this issue should be considered complete. 

- Upstream Vim is converting to a "new style" test format, and we don't want to duplicate work done there.
    - The "new style" tests are purported to run faster than the "old style".
- It is valuable for us to leave some fraction of the legacy tests in a state that is identical or nearly identical to Vim's tests, in order to make sure we are passing the same tests.
    - (Actually, we should probably set up a parallel build that runs the entire legacy test suite, though not per-PR, but at least on master)

We should move the legacy tests from `src/nvim/testdir` to `test/legacy`.
